src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2849,6 +2874,58 @@ FlowAnalysisAnnotations removeInapplicableNotNullWhenSense(ParameterSymbol param
             }
         }
 
+        private static TypeWithAnnotations ApplyLValueAnnotations(TypeWithAnnotations declaredType, FlowAnalysisAnnotations flowAnalysisAnnotations)
+        {
+            if ((flowAnalysisAnnotations & FlowAnalysisAnnotations.DisallowNull) != 0)
+            {
+                var typeSymbol = declaredType.Type;
+                if (declaredType.NullableAnnotation.IsNotAnnotated() || (typeSymbol.IsValueType && !typeSymbol.IsNullableType()))
+                {
+                    return declaredType;
+                }
+
+                return TypeWithAnnotations.Create(typeSymbol, NullableAnnotation.NotAnnotated, declaredType.CustomModifiers);
+            }
+            else if ((flowAnalysisAnnotations & FlowAnalysisAnnotations.AllowNull) != 0)
+            {
+                var typeSymbol = declaredType.Type;
+                if (declaredType.NullableAnnotation.IsAnnotated() || (typeSymbol.IsValueType && typeSymbol.IsNullableType()))
+                {
+                    return declaredType;
+                }
+
+                return TypeWithAnnotations.Create(typeSymbol, NullableAnnotation.Annotated, declaredType.CustomModifiers);
+            }
+
+            return declaredType;
+        }
+
+        private static TypeWithAnnotations ApplyRValueAnnotations(TypeWithAnnotations declaredType, FlowAnalysisAnnotations flowAnalysisAnnotations)
+        {
+            if ((flowAnalysisAnnotations & FlowAnalysisAnnotations.NotNull) == FlowAnalysisAnnotations.NotNull)
+            {
+                var typeSymbol = declaredType.Type;
+                if (declaredType.NullableAnnotation.IsNotAnnotated() || (typeSymbol.IsValueType && !typeSymbol.IsNullableType()))
+                {
+                    return declaredType;
+                }
+
+                return TypeWithAnnotations.Create(typeSymbol, NullableAnnotation.NotAnnotated, declaredType.CustomModifiers);
```

Comment:
Consider extracting helper methods to share code between this method and the one above. For instance, this could be `return AsNotAnnotated(declaredType);`